### PR TITLE
tanjiro: EMA weight averaging full 30-ep run on SOTA recipe

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**EMA weight averaging — full 30-epoch run on SOTA recipe** — EP4 screen (PR #1075) validated that the EMA mechanism works: EMA weights outperformed raw by 0.13–0.21pp across all metrics, and EMA crossed raw by EP3 showing the averaging is accumulating. The screen failed the EP4 gate only because 4 epochs is too short — EMA needs many low-noise averaging steps to compound into a metric-visible gain. This PR runs the full 30-epoch budget with the SOTA recipe so EMA has time to accumulate its full benefit.

### Background

PR #1075 (EP4 screen, W&B `9i1uu66t`) confirmed:
- EMA overtook raw model by EP3
- EMA stayed ahead through EP4 (EMA=7.2628% vs raw 7.4228%, delta −0.16pp)
- Consistent improvement across all metrics: vol_p −0.13pp, surf_p −0.13pp, WSS −0.16pp, τ_z −0.21pp
- Clean training — no NaN, no instability, healthy gradient norms

EMA is a regularization technique via weight-space averaging: the EMA model sits in a flatter loss landscape minimum than the raw checkpoint. In the context of WSS (a surface derivative quantity), smooth weight trajectories reduce the noise accumulated in the surface decoder's shear-field representation. The effect grows with more averaging steps — which is why 4 epochs showed promise but 30 will reveal the full benefit.

This is the **clean test**: same SOTA recipe as PR #972 (the single-model SOTA), identical flags, add only `--use-ema --ema-decay 0.9995 --ema-start-step 2000 --eval-raw-vs-ema`.

---

## Current Baselines (Corrected Dataset)

| Metric | Single-Model SOTA (PR #972) | Ensemble SOTA (PR #1064) | Target |
|--------|------------------------------|--------------------------|--------|
| val_abupt | 6.126% | 5.7758% | < 6.126% |
| test_abupt | 5.844% | 5.5199% | < 5.844% |
| test_vol_p | **3.643%** | **3.3630%** | ≤ 3.643% (hard floor) |
| test_surf_p | **3.577%** | — | ≤ 3.577% (hard floor) |
| test_WSS | 6.727% | 6.330% | < 5.85% (Transolver-3 target) |

**Single-model W&B:** `56bcqp3m` (train), `zxnhtagj` (eval) — PR #972

---

### Environment

```bash
cd target/
```

### Full 30-Epoch Training Command

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 \
  --use-sdf-importance-sampling --sdf-importance-alpha 0.25 \
  --surface-loss-weight 2.0 \
  --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --lr-cosine-t-max 30 --epochs 30 \
  --vol-points-schedule '0:16384:3:32768:6:49152:9:65536' \
  --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --validation-every 1 \
  --use-ema --ema-decay 0.9995 --ema-start-step 2000 --eval-raw-vs-ema \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --wandb-group tanjiro-ema-full-30ep \
  --wandb-name tanjiro/ema-full-30ep-sota-recipe
```

### Key flags

- `--use-ema`: Enable EMA weight averaging
- `--ema-decay 0.9995`: Higher decay than default 0.999 (half-life ~1386 steps vs ~693 — 2× longer memory)
- `--ema-start-step 2000`: Delay EMA accumulation until after LR warmup resolves; prevents noisy early weights polluting the average
- `--eval-raw-vs-ema`: **Critical** — evaluates both raw model AND EMA model at each validation epoch; report the EMA model's val_abupt

### What to Monitor

1. **EMA vs raw delta** — report both `val_primary/abupt_axis_mean_rel_l2_pct` (raw) and the EMA equivalent at each epoch. We expect the EMA advantage to grow from ~EP3 onward.
2. **val_WSS trajectory** — `val_primary/surface_wallshearstress_rel_l2_pct` for EMA vs raw. Primary interest is whether EMA reduces WSS.
3. **val_vol_p** — `val_primary/volume_pressure_rel_l2_pct` (EMA); hard floor is 3.643%.

### EP Gates (kill thresholds for early stopping only)

This is a full-budget run — kill only if clearly diverged:
- EP3: val_abupt > 8.0% → kill
- EP10: val_abupt > 7.0% → kill
- EP20: val_abupt > 6.5% → kill

### Kill-thresholds flag

```
--kill-thresholds "32925:val_primary/abupt_axis_mean_rel_l2_pct<8.0,109750:val_primary/abupt_axis_mean_rel_l2_pct<7.0,219500:val_primary/abupt_axis_mean_rel_l2_pct<6.5"
```

(Step counts: EP3≈32925 steps, EP10≈109750, EP20≈219500 at batch-size 4 on 34 val cases)

---

## Reporting

When the full run concludes (or if kill-threshold triggers), post results with terminal marker:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":<EMA_val_abupt>},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<EMA_test_abupt>}}
```

**Important:** The `primary_metric` should report the **EMA model** val_abupt (not raw). Include a table showing both EMA and raw metrics at the best EMA checkpoint epoch. Also report val_WSS and val_vol_p for both EMA and raw.

If killed early, set `"status":"killed"` and report the epoch at which kill occurred.

---

## Context: Why This Run Is Different from PR #1075

| | PR #1075 (EP4 screen) | This PR (full run) |
|-|------------------------|---------------------|
| Epochs | 4 | 30 |
| lr-cosine-t-max | 4 | 30 |
| vol-points-schedule | 0:16384:1:32768:2:49152:3:65536 | 0:16384:3:32768:6:49152:9:65536 |
| EMA start step | 2000 | 2000 |
| Result | EMA=7.26% (EP4 gate FAIL) | EMA benefit expected to compound over 30 epochs |

The mechanism is validated — we just need to let it run.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT tanjiro: Two blockers before launch on the current tay branch — need advisor clarification:

**1. SDF flags don't exist in tay train.py**

The command includes `--use-sdf-importance-sampling --sdf-importance-alpha 0.25`, but neither flag is in tay's Config dataclass (`train.py` line 75-141). The argparse will fail with 'unrecognized arguments' before training starts.

Searching the codebase:
- PR #972 (SOTA, `56bcqp3m`) used `--use-sdf-stratified-vol-sampling --sdf-stratified-alpha 2.0` (far-field weighting: `weight = 1 + α·|sdf|`) via a monkey-patch — never merged to tay.
- PR #1086 (EMA+SDF analog, `drivaerml-long-20260504` base) used `--use-sdf-stratified-vol-sampling --sdf-stratified-alpha 0.25` with the **inverse near-surface** formula `weight = 1 / (1 + α·|sdf|)` via a different monkey-patch.
- PR #1075 (EP4 EMA screen — direct predecessor of this PR) did **NOT** use any SDF flag.

Three possible interpretations:
  - (a) Drop SDF flags entirely, run the same recipe as PR #1075's screen (which already showed EMA crossover by EP3). Cleanest A/B vs the screen, but not 'same as #972'.
  - (b) Apply the PR #1086 inverse-near-surface monkey-patch with α=0.25 (matches the most recent SDF arm and the alpha in this PR).
  - (c) Apply the PR #972 far-field formula with α=0.25 (matches the SOTA recipe but with a much weaker α than #972's 2.0).

I lean toward **(b) PR #1086 monkey-patch (inverse near-surface, α=0.25)** because the α=0.25 matches and the inverse formula is what the current SDF-wave is using, but want your call before scope-creeping the experiment.

**2. EMA warm-start fix (PR #1087) is NOT in tay's trainer_runtime.py**

`EMA.update()` skips before `start_step` but never re-syncs the shadow from live weights when `step_counter == start_step`. With `--ema-start-step 2000` and `decay=0.9995`, the shadow will be biased toward init weights for the first few thousand updates after EMA activates (half-life ~1386 steps, so it takes ~5-10k steps to wash out). PR #1087 backported this fix to `drivaerml-long-20260504` (commit `15afb57`) but it wasn't merged to tay.

Should I cherry-pick `15afb57` as a bug-fix commit on this PR before the long run? Without it, the EMA shadow takes ~3-4 epochs to converge to a usable average, which would meaningfully blunt the 30-epoch comparison.

Standing by — will not launch until you confirm.

**Comment by morganmcg1** [comment]
ADVISOR: Both blockers resolved. Here's what to do:

**Blocker 1 — SDF flags: DROP THEM (option a)**

Do NOT use any SDF flag. The cleanest experiment is a direct extension of your EP4 screen (PR #1075) to 30 epochs — same recipe, just longer. Adding SDF is scope-creeping a different variable into what should be a clean EMA ablation. The PR #1075 screen showed EMA crossover by EP3, and that's what we want to confirm at full budget.

Remove `--use-sdf-importance-sampling --sdf-importance-alpha 0.25` from your launch command entirely.

**Blocker 2 — EMA warm-start fix: ALREADY APPLIED TO TAY**

I've committed the warm-start re-sync fix directly to the `tay` branch (backport of PR #1087). Your PR branch was created from tay, so you need to **rebase your branch on tay** to pick it up:

```bash
git fetch origin
git rebase origin/tay
```

After rebasing, verify the fix is present in `trainer_runtime.py`:
```bash
grep -A5 'step_counter == self.start_step' trainer_runtime.py
```
You should see the shadow re-sync block:
```python
        if self.step_counter == self.start_step:
            for name, param in model.named_parameters():
                if param.requires_grad and name in self.shadow:
                    self.shadow[name].copy_(param.detach())
            return
```

**Updated launch command (drop SDF flags):**

```bash
torchrun --standalone --nproc_per_node=8 train.py   --agent tanjiro   --optimizer lion --lr 9e-5 --weight-decay 5e-4   --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128   --batch-size 4   --pos-encoding-mode string_separable --use-qk-norm   --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0"   --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0   --surface-loss-weight 2.0   --grad-clip-norm 0.5 --lr-warmup-epochs 1   --lr-cosine-t-max 30 --epochs 30   --vol-points-schedule '0:16384:3:32768:6:49152:9:65536'   --no-compile-model   --train-surface-points 65536 --eval-surface-points 65536   --train-volume-points 65536 --eval-volume-points 65536   --validation-every 1   --use-ema --ema-decay 0.9995 --ema-start-step 2000 --eval-raw-vs-ema   --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511   --wandb-group tanjiro-ema-full-30ep   --wandb-name tanjiro/ema-full-30ep-sota-recipe   --kill-thresholds "32925:val_primary/abupt_axis_mean_rel_l2_pct<8.0,109750:val_primary/abupt_axis_mean_rel_l2_pct<7.0,219500:val_primary/abupt_axis_mean_rel_l2_pct<6.5"
```

Rebase, verify the fix, then launch. Good luck!

**Comment by morganmcg1** [comment]
STUDENT tanjiro: Rebased on tay (commit 909526a), EMA warm-start fix verified present in trainer_runtime.py. Launched full 30-epoch run with SDF flags dropped per advisor instruction.

**Run launched:** `xn4gfjc9` (rank0) — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/xn4gfjc9

**Throughput estimate** (based on early steps and PR #1075 screen):
- ep0-2 (vol=16384): ~77 min/ep × 3 = 231 min
- ep3-5 (vol=32768): ~46 min/ep × 3 = 138 min
- ep6-8 (vol=49152): ~36 min/ep × 3 = 108 min
- ep9-29 (vol=65536): ~30 min/ep × 21 = 630 min
- **Total expected: ~18h** (well beyond the 6h student session timeout, but training runs persist across entrypoint re-invocations)

Will monitor via sparse wakeups and post results when complete.
